mypy/checker.py

```diff
@@ -5193,6 +5234,160 @@ def refine_identity_comparison_expression(self,
+        the output TypeMaps.
+
+        Although this function could theoretically try setting the types of the operands
+        in the chains to the meet, doing that causes too many issues in real-world code.
+        Instead, we use 'is_valid_target' to identify which of the given chain types
+        we could plausibly use as the refined type for the expressions in the chain.
+
+        Similarly, 'coerce_only_in_literal_context' controls whether we should try coercing
+        expressions in the chain to a Literal type. Performing this coercion is sometimes
+        too aggressive of a narrowing, depending on context.
+        """
+
+        target: Optional[int] = None
+        target_index: Optional[int] = None
+        possible_target_indices = []
+        for i in chain_indices:
+            expr_type = operand_types[i]
+            expr_type = coerce_to_literal(expr_type)
+            proper_type = get_proper_type(expr_type)
+            if not isinstance(proper_type, LiteralType):
+                continue
+            if target and target != proper_type.value:
+                if operator in {'==', '!='}:
+                    # We have multiple different target values. So the 'if' branch
+                    # must be unreachable.
+                    return None, {}
+                else:
+                    # Other operators can go either way
+                    return {}, {}
+
+            target = proper_type.value  # type: ignore[assignment]
+            target_index = i
+            possible_target_indices.append(i)
+
+        # There's nothing we can currently infer if none of the operands are valid targets,
+        # so we end early and infer nothing.
+        if target is None:
+            return {}, {}
+
+        partial_type_maps = []
+        for i in chain_indices:
+            # Naturally, we can't refine operands which are not permitted to be refined.
+            if i not in narrowable_operand_indices:
+                continue
+
+            # we already checked that operand[i] is CallExpr since it is narrowable
+            expr = operands[i].args[0]  # type: ignore[attr-defined]
+            expr_type = self.type_map[expr]
+
+            # We intentionally use 'conditional_type_map' directly here instead of
+            # 'self.conditional_type_map_with_intersection': we only compute ad-hoc
+            # intersections when working with pure instances.
+            partial_type_maps.append(
+                self.conditional_len_map(operator, expr, expr_type, i, target, target_index))
+
+        return reduce_conditional_maps(partial_type_maps)
+
+    def narrow_type_by_length(self, operator: str, typ: Type, length: int) -> Type:
+        if operator not in {"==", "!="}:
+            return typ
+        proper_type = get_proper_type(typ)
+        if (isinstance(proper_type, Instance) and proper_type.type.fullname == "builtins.tuple"
+                and length >= 0):
```

Comment:
I notice this excludes zero-length tuples, and it doesn't seem like they're tested. We might want a special case for them, though it's not too likely in code that you'd do anything with the result.